Docs update (#515)
fiberplane/fiberplane: www/src/content/docs/docs/get-started.mdx

---
title: Get Started
description: How to get started with Fiberplane Playground
---

import PackageManagers from "@/components/PackageManagers.astro";
import Platforms from "@/components/Platforms.astro";
import { Aside, Steps, LinkCard } from "@astrojs/starlight/components";

> Fiberplane is a [Hono](https://hono.dev) middleware that provides a built-in Playground UI for exploring and testing Hono service APIs. It leverages [OpenAPI](https://www.openapis.org/what-is-openapi) to deliver interactive API documentation, allowing users to send requests with generated payloads through the UI.

Reviewer: ```suggestion
Fiberplane is an embeddable, interactive playground for exploring and testing [Hono](https://hono.dev) APIs. It leverages [OpenAPI](https://www.openapis.org/what-is-openapi) to render rich API documentation and enable developers to send test requests through its UI.
```

### 1. Set up a Hono project

You can set up a basic Hono project via your terminal:

<PackageManagers type="create" pkg="hono@latest" args="my-app" />

> If you want a project that includes an ORM, a database, the option to include Hono Zod OpenAPI (see step 2), Fiberplane middleware (see step 3) and is ready to deploy as a Cloudflare Worker, we recommend using [HONC 🪿](https://honc.dev/):

Reviewer: the parenthetical steps are a little overwhelming (e.g., `(step 2)`)

```suggestion
If you want a project that includes a database, integration with Hono Zod OpenAPI, and is ready to deploy as a Cloudflare Worker, we recommend using [HONC 🪿](https://honc.dev/):
```

<PackageManagers type="create" pkg="honc-app@latest" />

### 2. Add an OpenAPI spec to Hono

> The Fiberplane Playground Hono middleware works with an OpenAPI specification. There are two packages that help you define OpenAPI specs in your Hono application.

Reviewer: ```suggestion
The Fiberplane Playground requires an OpenAPI spec to work. There are two packages that help you define an OpenAPI spec for your Hono application:
```

> [Zod OpenAPI Hono](https://github.com/honojs/middleware/tree/main/packages/zod-openapi) helps you design OpenAPI specs from the start, ensuring all definitions align with the extended OpenAPI class.

Reviewer: ```suggestion
[Zod OpenAPI Hono](https://github.com/honojs/middleware/tree/main/packages/zod-openapi) helps you design an OpenAPI spec from the start, ensuring type safety between your code and the spec definitions.
```

> [Hono OpenAPI](https://hono.dev/examples/hono-openapi) generates specs from existing validation schemas, allowing for incremental spec building.

Reviewer: ```suggestion
[Hono OpenAPI](https://hono.dev/examples/hono-openapi) generates a spec from your existing validation schemas, allowing for incremental adoption of OpenAPI for your api.
```

<Aside type="note">
  For fine-grained and detailed OpenAPI specs, we recommend using these packages
  for accuracy and maintainability, though you can also generate a spec from
  existing code with Fiberplane's `createOpenAPISpec` helper function.
</Aside>

<details>

<summary>Zod OpenAPI Hono</summary>

> [Zod OpenAPI Hono](https://github.com/honojs/middleware/tree/main/packages/zod-openapi) provides input validation, type safety, and OpenAPI documentation.

Reviewer: we can keep for now, but this might be excessively detailed for a "Getting started" docs page. we could always point people to the README or docs for this package, or set up own own guide

Author: sounds good... 

I think we discussed in the initial docs meeting that we want to have guides for all three OpenAPI spec options in our own docs with examples. But we can either make them guides or link to the existing docs.

> It extends the Hono class and routes have to be defined and passed as parameters in the handlers.

Reviewer: ```suggestion
It extends the Hono class, and requires that you pass a specification of your routes before defining the logic of your handlers. This makes it possible to type-check your api inputs and outputs against your api spec.
```

> Install the package alongside hono and zod:

Reviewer: ```suggestion
Install the package alongside Hono and Zod:
```

const app = new OpenAPIHono<{
  Bindings: Bindings;
  Variables: Variables;
}>();
```

Define your input data schemas in Zod:

```ts
const UserResponseSchema = z
  .object({
    id: z.number().int().openapi({
      description: "The unique identifier for the user",
      example: 1
    }),

    name: z.string().min(1).max(255).openapi({
      description: "The user's full name",
      example: "Matthew"
    }),

    email: z.string().email().openapi({
      description: "The user's email address",
      example: "matthew@cloudflare.com"
    })
  })
  .openapi("User", {
    description: "Schema for user response including ID"
  });

const CreateUserSchema = UserResponseSchema.omit({ id: true }).openapi(
  "CreateUser",
  {
    description: "Schema for creating a new user"
  }
);
```

<Aside type="note">
  `.extend` does not propagate the openapi metadata to the new schema. Instead
  use `.merge` .
</Aside>

Define your routes in Zod and include the input and output data schemas if needed:

```ts
const createUser = createRoute({
  method: "post",
  path: "/api/user",
  request: {
    // Validate request body using Zod schemas
    body: {

>       required: true, // NOTE: this is important to set to true, otherwise the route will accept empty body

Reviewer: ```suggestion
     // NOTE: this is important to set to true, otherwise the route will accept empty bodies and invalid content types
      required: true,
```

      description: "The user's full name",
      example: "Matthew"
    }),

    email: z.string().email().openapi({
      description: "The user's email address",
      example: "matthew@cloudflare.com"
    })
  })
  .openapi({
    ref: "User",
    description: "Schema for user response including ID"
  });

const CreateUserSchema = UserResponseSchema.omit({ id: true }).openapi({
  ref: "CreateUser",
  description: "Schema for creating a new user"
});
```

<Aside type="note">
  `.extend` does not propagate the openapi metadata to the new schema. Instead
  use `.merge` .
</Aside>
Next add a description to the route:

```ts
import { describeRoute } from "hono-openapi";
import { resolver, validator as zValidator } from "hono-openapi/zod";

app.post(
  "/api/user",
  describeRoute({
    description: "Create a new user",
    requestBody: {
      content: {
        "application/json": {
          schema: resolver(CreateUserSchema)
        }
      }
    },
    responses: {
      201: {
        description: "The created user",
        content: {
          "application/json": {
            schema: resolver(UserResponseSchema)
          }
        }
      }
    }
  }),

>   zValidator("json", CreateUserSchema),

Reviewer: ```suggestion
  // Middleware to validate incoming request bodies
  zValidator("json", CreateUserSchema),
```

      info: {
        title: "Hono",
        version: "1.0.0",
        description: "API for users"
      }
    },
    servers: [
      {
        url: "http://localhost:8787",
        description: "Local server"
      }
    ]
  })
);
```

</details>

<details>
<summary>Fiberplane's `createOpenAPISpec` helper function</summary>

The other option is to integrate OpenAPI documentation into an existing Hono codebase. Refactoring the application to use Hono Zod OpenAPI or Rhinobase provides fine-grained documentation, offering long-term benefits, but requires more effort. For a quicker solution, Fiberplane provides a helper function to generate an OpenAPI spec from the existing code.

{" "}
<PackageManagers type="add" pkg="@fiberplane/hono" />

This step involves importing the `createOpenAPISpec` function from Fiberplane and creating an endpoint at `/openapi.json` to generate and return an OpenAPI spec.

```ts
import { createOpenAPISpec } from "@fiberplane/hono";

app.get("/openapi.json", (c) => {
  const spec = createOpenAPISpec(app, {
    info: { title: "My API", version: "1.0.0" }
  });
  return c.json(spec);
});
```

</details>

### 3. Add Fiberplane Middleware

> With the OpenAPI spec available, the Fiberplane middleware can now be installed and pointed to the OpenAPI spec.

Reviewer: reworded to avoid passive voice

```suggestion
Now that the OpenAPI spec is available, mount the Fiberplane middleware and point it to the OpenAPI spec.
```

<PackageManagers type="add" pkg="@fiberplane/hono" />

```ts
import { createFiberplane } from "@fiberplane/hono";

app.use(
  "/fp/*",
  createFiberplane({
    openapi: {
      url: "/openapi.json"
    }
  })
);
```

<Aside type="note">

>   It is important to mount Fiberplane's middleware last in the codebase.

Reviewer: ```suggestion
  It is important to mount Fiberplane's middleware _after_ all of your route definitions.
```
